gammapy/detect/cwt.py

```diff
@@ -57,200 +62,552 @@ def difference_of_gauss_kernel(radius, scale_step, n_sigmas=8):
+        answer : boolean
+            Answer if CWT has converged.
+        """
+        answer = False
+        residual = data.counts - (data.model + data.approx)
+        variance = residual.var()
+        if self.previous_variance is not None:
+            variance_ratio = abs((self.previous_variance - variance) / self.previous_variance)
+            answer = variance_ratio < self.tol
+            log.info('Variance ratio: {:.7f}'.format(variance_ratio))
+        self.previous_variance = variance
+        log.info('')
+        return answer
+
+    def analyze(self, data):
+        """
+        Run iterative filter peak algorithm. The algorithm modifies the original data.
 
-    - max_scale_image : 2D image, value = index of scale where emission is most significant
+        Parameters
+        ----------
+        data : `~gammapy.detect.CWTData`
+            Input images.
+        """
+        if self.history is not None:
+            self.history = [copy.deepcopy(data)]
+
+        for n_iter in range(self.max_iter):
+            log.info('Start iteration {0}'.format(n_iter + 1))
+            self._execute_iteration(data=data)
+            if self.history is not None:
+                self.history.append(copy.deepcopy(data))
+            if self._is_converged(data=data):
+                break
+
+        if n_iter != self.max_iter:
+            log.info('Convergence reached at iteration {0}'.format(n_iter + 1))
+        else:
+            log.info('Convergence not formally reached at iteration {0}'.format(n_iter + 1))
+
+    def get_history(self):
+        """
+        Return cwt data from all the iterations.
+
+        Returns
+        -------
+        data : list
+            List of '~gamapy.detect.CWTData' objects.
+        """
+        if self.history is None:
+            raise AttributeError('This CWT object does not support saving cwt data images.')
+        else:
+            return copy.deepcopy(self.history)
+
+
+class CWTKernels(object):
+    """
+    Conduct arrays of kernels and scales for CWT algorithm.
+
+    Parameters
+    ----------
+    n_scale : int
+        Number of scales.
+    min_scale : double
+        First scale used.
+    step_scale : double
+        Base scaling factor.
+    old : boolean (default True)
+        DEBUG attribute. If False, use astropy kernels.
+
+    Attributes
+    ----------
+    n_scale : int
+        Number of scales considered.
+    min_scale : double
+        First scale used.
+    step_scale : double
+        Base scaling factor.
+    scales : '~numpy.ndarray'
+        Grid of scales.
+    kern_base : dict
+        Dictionary of scale powers as keys and 2D kernel arrays.
+        (mexican hat) as values
+    kern_approx : '~numpy.ndarray'
+        2D Gaussian kernel array from maximum scale.
+
+    Examples
+    --------
+    >>> from gammapy.detect import CWTKernels
+    >>> kernels = CWTKernels(n_scale=2, min_scale=6.0, step_scale=1.3)
+    >>> print (kernels.scales)
+    [ 6.   7.8]
```

Comment:
Can you maybe choose a slightly more interesting example where you have 3 or 4 scales,
and then show the output of `kernels.info` here?